src/services/virus-scanner.ts

```diff
@@ -0,0 +1,133 @@
+
+// Handles file upload and virus scanning. Expects a single file in the request, streams it to the ClamAV scanner,
+// and writes the file to a temporary location. If the file is infected, it throws an error and deletes the file.
+// Note: This function assumes that the request has been processed by a middleware that populates req.files with the
+// uploaded files streams. If we want to handle multiple files in a single request, we would need to iterate over
+// req.files, see https://github.com/rafasofizada/pechkin/blob/master/examples/express.js
+export const uploadAvScan = async (req: Request): Promise<TempFile> => {
+  const start = performance.now();
+  const iterable = ((await req.files?.next()) as FileIterator) || {};
+  const { stream, filename, mimeType } = iterable.value || {};
+
+  if (!stream || !filename || !mimeType) {
+    throw new BadRequestException('errors.file_upload.missing');
+  }
+
+  let virusScanner: Transform;
+  let tmpFileStream: WriteStream;
+  let tmpFile: TempFile = { path: '', originalname: filename, mimetype: mimeType };
+
+  try {
+    virusScanner = await getAVPassthrough();
+    tmpFileStream = getTmpFileStream(tmpFile, (finished: TempFile) => (tmpFile = finished));
+  } catch (err) {
+    logger.error(err, 'There was a problem initializing the virus scanner or temporary file stream');
+    throw new UnknownException('errors.file_upload.stream_failure');
+  }
+
+  // pipeline will not error on infected files, need to wait for 'scan-complete' event to know the result
+  // it will wait for the file to finish writing before resolving the promise though
+  await pipeline(stream, virusScanner, tmpFileStream).catch((err: unknown) => {
+    logger.error(err, 'There was a problem streaming the file upload');
+    throw new UnknownException('errors.file_upload.stream_failure');
+  });
+
+  if (env === AppEnv.Ci) {
+    logger.info(`Skipping AV scan in CI environment for file "${filename}"`);
+    return tmpFile;
+  }
+
+  // wait for the scan to complete before returning the temporary file
+  return new Promise((resolve, reject) => {
+    virusScanner.on('timeout', () => reject(new UnknownException('errors.file_upload.av_timeout')));
+
+    virusScanner.on('error', (err) => {
+      logger.error(err, 'There was a problem with the virus scanner');
+      reject(new UnknownException('errors.file_upload.scan_failure'));
+    });
+
+    virusScanner.on('scan-complete', async (result) => {
+      const time = Math.round(performance.now() - start);
+
+      if (result.isInfected) {
+        cleanupTmpFile(tmpFile);
+        const viruses = result.viruses.join(', ');
+        logger.warn(`AV Scan complete. File "${filename}" is infected with: "${viruses}", time: ${time}ms`);
+        reject(new BadRequestException('errors.file_upload.infected'));
+        return;
+      }
+
+      logger.info(`AV Scan complete. File "${filename}" is clean, time: ${time}ms`);
+      resolve(tmpFile);
+    });
+  });
+};
```

Comment:
This function returns a TempFile object that provides info (path, etc) about the uploaded file location.

We could optionally create a separate AV function that allows us to pass in stream handlers (e.g. a CSV parser and a Transform that writes to Postgres?) instead of writing to a temporary file, but that's a much bigger chunk of work.